src/generate.py

```diff
@@ -780,576 +660,153 @@ def flatten(tree, structs, visited):
-
-yajl_val get_val (yajl_val tree, const char *name, yajl_type type);
-
-string_cells *read_map_string_string (yajl_val src);
-
-void free_cells (string_cells *cells);
-
-void *safe_malloc (size_t size);
-
-#endif
-""")
-
-def generate_common_C_code(c_file):
-    c_file.write("""/* autogenerated file */
-#include "json_common.h"
-
-bool reformat_integer (void *ctx, long long int num) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_integer (g,num));
-}
-
-bool reformat_double (void *ctx, double num) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_double (g,num));
-}
-
-bool reformat_number (void *ctx, const char *str, size_t len) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_number (g, str, len));
-}
-
-bool reformat_string (void *ctx, const unsigned char *str, size_t len) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_string (g, str, len));
-}
-
-bool reformat_null (void *ctx) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_null (g));
-}
-
-bool reformat_boolean (void *ctx, int boolean) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_bool (g, boolean));
-}
-
-bool reformat_map_key (void *ctx, const unsigned char *str, size_t len) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_string (g, str, len));
-}
-
-bool reformat_start_map (void *ctx) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_map_open (g));
-}
-
-bool reformat_end_map (void *ctx) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_map_close (g));
-}
-
-bool reformat_start_array (void *ctx) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_array_open (g));
-}
-
-bool reformat_end_array (void *ctx) {
-    yajl_gen g = (yajl_gen) ctx;
-    GEN_AND_RETURN (yajl_gen_array_close (g));
-}
-
-bool gen_map_string_string (void *ctx, string_cells *cells) {
-    bool stat = true;
-    yajl_gen g = (yajl_gen) ctx;
-    size_t i = 0;
-    if (!cells)
-        return false;\n
-    stat = reformat_start_map (g);
-    if (!stat)
-        return false;\n
-    for (i = 0; i < cells->len; i++) {
-        stat = reformat_map_key (g, (unsigned char *) cells->keys[i], strlen (cells->keys[i]));
-        if (!stat)
-            return false;
-        stat = reformat_string (g, (unsigned char *) cells->values[i], strlen (cells->values[i]));
-        if (!stat)
-            return false;
-    }\n
-    stat = reformat_end_map (g);
-    if (!stat)
-        return false;
-    return true;
-}
-
-bool json_gen_init (yajl_gen *g) {
-    *g = yajl_gen_alloc (NULL);
-    if (NULL == *g)
-        return false;\n
-    yajl_gen_config (*g, yajl_gen_beautify, 1);
-    yajl_gen_config (*g, yajl_gen_validate_utf8, 1);
-    return true;
-}
-
-yajl_val get_val (yajl_val tree, const char *name, yajl_type type) {
-    const char *path[] = { name, NULL };
-    return yajl_tree_get (tree, path, type);
-}
-
-void free_cells (string_cells *cells) {
```

Comment:
looks like at least the definition of free_cells is missing.

As well as the definition of `string_cells`.

Reply:
@giuseppe I build crun with the modifed libocispec successfully. And I pushed a PR to the crun project.  The function `free_cells` and `string_cells` were used by mapStringString, but we refact the implement for mapStringString, they are useless now.
![image](https://user-images.githubusercontent.com/26912745/72657387-9eebd500-39de-11ea-9087-fa1f2e788696.png)